cloudinit/net/dhcp.py

```diff
@@ -473,6 +473,7 @@ def get_dhclient_d():
             "/var/lib/dhclient",
             "/var/lib/dhcp",
             "/var/lib/NetworkManager",
+            "/var/db",
```

Comment:
I'm on the fence about whether to make this a (free)bsd check before adding. On one hand, I check a couple of linuxen, and some have the directory, others don't. Why check this directory on those distros? On the other hand, there were few files in that directory and only a single `listdir` called, so it shouldn't be "slow".

This function is one that I've previously considered breaking out into `distros/` anyways, since I think `/var/lib/dhcp/` and `/var/lib/dhclient/` are distro-specific iirc.

Thoughts?

Reply:
I briefly checked the `distros/` directory: seems its purpose is to isolate distro-specific code in there. So this means that the cloud-init has a reliable way to detect which os/distro it is running on? 

If yes, then its indeed better to implement this function per-distro.
Perhaps implement a method in `distros.Distro` that uses a class attribute [such as this](https://github.com/canonical/cloud-init/blob/main/cloudinit/distros/debian.py#L34) and override that class attribute per child distro class.

Reply:
> Perhaps implement a method in distros.Distro that uses a class attribute with the dhclient directory [such as this](https://github.com/canonical/cloud-init/blob/main/cloudinit/distros/debian.py#L34) and override that class attribute per child distro class.

@phsm yes, this sounds like the right direction. I think we have a reference to the distro instance higher up that we'll just need to pass down the stack as an argument to access it in this function.

Reply:
I just made another commit with an example how I see it done:
- main Distro class has two attributes: dhclient_lease_directory, dhclient_lease_file_regex defining the lease directory for lease files in a distro, and a regex helping to find those lease files in there.
- every class of a distro that uses dhclient has those 2 attributes defined.
- Cloudstack's get_vr_address() now detects a distro and supplies dhcp.IscDhclient.get_latest_lease() with the lease_dir an a regex.
- there is no need for get_dhclient_d() anymore
- Although get_latest_lease() function signature is changed, I don't see it being used anywhere except Cloudstack DS so it shouldn't break anything.

The purpose of this commit is to gather your feedback if i'm working in the right direction. 
Also, do I detect the distro correctly here?

Reply:
I just updated the PR with the following:
1. Updated the distros files with dhclient lease director/regex for distros that support DHClient. I installed all of them on a virtual machine to verify the paths:

- CentOS: updated centos.py to contain dhclient parameters for Centos 7. I assume no one uses CentOS > 7 on servers because they're EOL.
- RHEL: updated rhel.py with dhclient parameters. To use dhclient in RHEL, `dhcp=dhclient` parameter must be set in NetworkManager config.
- AlmaLinux: identical to RHEL.
- Rocky: identical to RHEL.
- Debian: updated debian.py with dhclient parameters.
- Alpine: doesn't use dhclient, uses udhcpc.
- Amazon Linux: updated amazon.py with dhclient parameters.
- CloudLinux: identical to RHEL.
- EuroLinux: identical to RHEL.
- Fedora: identical to RHEL
- FreeBSD: updated freebsd.py with dhclient parameters.
- Gentoo: doesn't use dhcpcd, uses dhcpcd.
- Mariner: uses systemd-networkd.
- Miracle: identical to RHEL
- Open Cloud OS: I couldn't even download it an ISO image to test. I assume its identical to RHEL.
- OpenEuler: identical to RHEL
- OpenSUSE: doesn't use dhclient, uses wicked.
- SLES: I don't have a copy of SLES, I assume it also uses wicked.
- Virtuozzo: I don't have a copy of it, I assume its identical to RHEL.
- Tencentos: didn't test, assume its identical to RHEL. 

2. Tests: 
- Added tests for finding a lease in different distros that can use dhclient.
- Removed `TestGetLatestLease` testcase from `test_cloudstack.py` in its entirety: Cloudstack tests shouldn't test obtaining DHCP lease, its already covered in `test_dhcp.py`

Reply:
Thanks @phsm for the update. I think that there is a missing detail here.

> doesn't use dhclient

Use of wicked / networkd / networkmanager / etc doesn't necessarily mean that dhclient isn't used. In fact, cloud-init itself uses dhclient during the [local](https://cloudinit.readthedocs.io/en/latest/explanation/boot.html#local) stage. Cloud-init sets up an ephemeral network so that it can reach out to the IMDS to get configuration. This allows cloud-init to write the networking configuration locally *before* the distro's built-in network manager is even started.

You've done some good work on this refactor, but I think that by default all distros should probably be checking for the isc-dhclient path, at least until `udhcpc` is fully featured for alpine, etc (with classless static routes and unknown dhcp option support) and dhcpcd replaces isc-dhclient on other distros.